### PR TITLE
Mv timed grooming

### PR DESCRIPTION
**Pull request by matthewvon**
This branch initiates smaller compaction sets in levels 0 and 1 based upon elapsed time.  This is to improve read latencies in datasets with small to medium write loads.  

See https://github.com/basho/leveldb/wiki/mv-timed-grooming for details.

### CODE REVIEW

File: db/dbformat.h [hunk 1]
```diff
@@ -31,6 +31,8 @@ static const size_t kL0_CompactionTrigger = 6;
 // Level-0 (any overlapped level) number of files where a grooming
 //     compaction could start
 static const size_t kL0_GroomingTrigger = 4;
+static const size_t kL0_GroomingTrigger10min = 2;
```
Comment (paulplace) [inline db/dbformat.h:34]: Can remove "static" from all these constants

File: db/version_set.cc [hunk 2]
```diff
@@ -1125,6 +1128,22 @@ VersionSet::Finalize(Version* v)
         // consider this level
         if (compact_ok)
         {
+            size_t grooming_trigger;
+
+            // which grooming trigger point?  based upon how long
+            //  since last compaction on this level
+            //   - less than 10 minutes?
+            if (micros_now<(m_CompactionStatus[level].m_LastCompaction + (1000000 * 60 *10)))
```
Comment (paulplace) [inline db/version_set.cc]: I recommend pre-computing the elapsed time as "uint64_t elapsed_micros = micros_now - m_CompactionStatus[level].m_LastCompaction;" so that the comparisons become:

```
uint64_t elapsed_micros = micros_now - m_CompactionStatus[level].m_LastCompaction;
if (elapsed_micros < (1000000 * 60 *10))
    ...
else if (elapsed_micros < (1000000 * 60 *20))
```

File: db/version_set.cc [hunk 3]
```diff
@@ -1125,6 +1128,22 @@ VersionSet::Finalize(Version* v)
...
+                grooming_trigger=config::kL0_GroomingTrigger;
+
+            //   - less than 20 minutes?
+            else if (micros_now<(m_CompactionStatus[level].m_LastCompaction + (1000000 * 60 *20)))
```
Comment (paulplace) [inline db/version_set.cc]: I recommend defining (1000000 \* 60 *10) and (1000000 \* 60 *20) as constants (probably in db/dbformat.h, along with kL0_GroomingTrigger10min and kL0_GroomingTrigger20min)

File: db/version_set.h [hunk 4]
```diff
@@ -374,9 +375,10 @@ class VersionSet {
   {
       bool m_Submitted;     //!< level submitted to hot thread pool
       bool m_Running;       //!< thread actually running compaction
+      uint64_t m_LastCompaction; //!<NowMicros() when last compaction completed
 
       CompactionStatus_s()
-      : m_Submitted(false), m_Running(false)
+      : m_Submitted(false), m_Running(false), m_LastCompaction(Env::Default()->NowMicros())
```
Comment (paulplace) [inline db/version_set.h]: The problem with initializing m_LastCompaction via Env::Default()->NowMicros() is that you're mixing and matching environment objects, since m_LastCompaction is set to env_->NowMicros() in SetCompactionDone() above. Normally this assumption is fine, but a different environment object might be used in unit tests. Should this be initialized to 0 here?

File: db/dbformat.h (hunk 1 above)
Comment (matthewvon) [inline db/dbformat.h:34]: I do not disagree.  But "static const" is part of the original Google writing.  I see no benefit to removing.  Is there a benefit beyond removing redundancy of declaration?

File: db/version_set.cc (hunk 2 above)
Comment (matthewvon) [inline db/version_set.cc]: I like the thought.  The thing that bothers me is that clock is not guaranteed to move forward on platforms forced to use gettimeofday().  So micros_now could be smaller than m_LastCompaction.  Going to add test for the case and force elapsed to zero.

File: db/version_set.h (hunk 4 above)
Comment (matthewvon) [inline db/version_set.h]: Ok, initializing to zero.  Change test in version_set.cc to explicitly check for value of zero and do right thing.

File: db/dbformat.h (hunk 1 above)
Comment (paulplace) [inline db/dbformat.h:34]: It depends on the compiler and linker, but I have seen static variables that are declared and instantiated in a header file cause an instance of the variable to be statically instantiated in each source file that includes the header file. It's not a big deal since the variables are also const, but it does add extra space by duplicating the variables in multiple source files.

In this case, I am cool with not changing anything.

Comment (matthewvon) [inline db/dbformat.h:34]: Yeah, let this one slide.

File: db/version_set.h (hunk 4 above)
Comment (paulplace) [inline db/version_set.h]: Hmm...I think the "initialize to 0" is in the wrong place. Should be in the ctor for CompactionStatus_s. With this change, SetCompactionDone() will always set m_LastCompaction to 0, which means the calculation in VerstionSet::Finalize() will always set elapsed_micros to 0.

Comment (matthewvon) [inline db/version_set.h]: oops

**Comment by paulplace** [comment]
+1 code review looks good
